Codes/Clean Data/getHasbrouck.R

```diff
@@ -0,0 +1,108 @@
+function(indexExe, lags, price, direction, sz) {
+  #vector autoregressive (VAR) system of differences in log prices ($r_t=\Delta p_t$) and trade-related characteristics $x_t$:
+  
+  y <- diff(log(price[indexExe]))
+  
+  #The column vector of trade variables $\mathbf{x}_t$ includes:
+  #(1) a sign indicator reflecting the direction of the trade,
+  #(2) signed trading volume, and
+  #(3) the signed square root of trading volume.
+  
+  x1 <-
+    direction[indexExe] * (-1)
+  x1 <-
+    x1[-1] #note that direction needs to be multiplied by -1, as it refers to sign on submission (not execution)
+  x2 <- sz[indexExe]
+  x2 <- x2[-1] * x1
+  x3 <- sz[indexExe]
+  x3 <- sqrt(x3[-1]) * x1
+  
+  ##step one: get residuals from VAR
+  model <- cbind(y, x1, x2, x3)
+  dimx <- ncol(model)
+  test <- try(VAR(model,
+                  p = lags,
+                  output = F,
+                  include.mean = F), silent = T)
+  if (class(test) == "try-error") {
+    return(c(NA, NA, NA))
+    break
+  } #break if estimation is not possible (e.g., singular matrix)
+  
+  res <- residuals(test)
+  rm(test)
+  
+  ##step two: estimate VMA parameters using residuals from VAR
+  T <- nrow(res)
+  ist <- lags + 1
+  
+  xmtx <- NULL
+  for (j in 0:lags) {
+    xmtx <- cbind(xmtx, res[(ist - j):(T - j), ])
+  }
+  y <- y[-(1:(2 * lags))]
+  label <- c("V1.L", "V21.L", "V22.L", "V23.L")
+  colnames(xmtx) <-
+    c(
+      paste0(label, 0),
+      paste0(label, 1),
+      paste0(label, 2),
+      paste0(label, 3),
+      paste0(label, 4),
+      paste0(label, 5)
+    )
+  
+  fit <- try(lm(y ~ xmtx - 1), silent = T)

```

Comment:
I understand right, $a_j^*$ and $b_j^*$ are deterministic functions of $a_j$, $b_j$, $c_j$ and $d_j$. The linear regression is a trick to compute those values. Am I right?

Reply:
st (pricing errors) is a function of alpha and beta, and alpha and beta are the functions of coefficients from a VMA representation of a VAR system. The errors from the VAR are used to approximate the coefficients of the VMA. I'm not sure if this is the best way to do this, but it might be the simplest. The VAR errors are sometimes used as the starting values when estimating a VMA (see, e.g., https://faculty.chicagobooth.edu/ruey.tsay/teaching/mts/sp2011/VMA1.R or https://communities.sas.com/t5/Statistical-Procedures/R-Square-as-in-Hasbrouck-s-Review-of-Financial-Studies-1991/td-p/44948 for an example in SAS) but a few early attempts at estimating the VMA lead me to some convergence troubles. I figured that this "simpler" approach would also be faster.